### PR TITLE
REF/PERF: MultiIndex.get_locs to use boolean arrays internally

### PR DESCRIPTION
**Pull request by lukemanley**
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Use boolean arrays internally within `MultiIndex.get_locs` rather than int64 indexes. Logical operations show performance improvements over intersecting int64 indexes. The output remains an integer positional indexer. 

```
       before           after         ratio
     [17dda440]       [94121581]
     <main>           <multiindex-get-locs-bool-arrays>
-        563±10μs          519±8μs     0.92  indexing.MultiIndexing.time_loc_all_scalars(True)
-      33.9±0.6ms       30.4±0.4ms     0.89  indexing.MultiIndexing.time_loc_all_null_slices(True)
-        38.7±1ms         34.5±1ms     0.89  indexing.MultiIndexing.time_loc_all_null_slices(False)
-     1.62±0.02ms      1.43±0.01ms     0.88  indexing.MultiIndexing.time_loc_all_slices(True)
-     6.40±0.06ms       5.53±0.1ms     0.86  indexing.MultiIndexing.time_loc_all_bool_indexers(True)
-         107±1ms       41.6±0.6ms     0.39  indexing.MultiIndexing.time_loc_all_lists(True)
-      34.6±0.8ms       8.24±0.4ms     0.24  indexing.MultiIndexing.time_loc_all_slices(False)
-         236±4ms       23.6±0.2ms     0.10  indexing.MultiIndexing.time_loc_all_lists(False)
-      97.3±0.7ms       9.16±0.4ms     0.09  indexing.MultiIndexing.time_loc_null_slice_plus_slice(False)
-      36.5±0.5ms      1.36±0.03ms     0.04  indexing.MultiIndexing.time_loc_null_slice_plus_slice(True)
```

### CODE REVIEW

File: pandas/core/indexes/multi.py [hunk 1]
```diff
@@ -3148,50 +3147,41 @@ def maybe_mi_droplevels(indexer, levels):
...
 
         def convert_indexer(start, stop, step, indexer=indexer, codes=level_codes):
-            # given the inputs and the codes/indexer, compute an indexer set
+            # given the inputs and the codes/indexer, compute a bool indexer
             # if we have a provided indexer, then this need not consider
             # the entire labels set
             if step is not None and step < 0:
                 # Switch elements for negative step size
                 start, stop = stop - 1, start - 1
-            r = np.arange(start, stop, step)
```
Comment (jreback) [inline pandas/core/indexes/multi.py:3168]: can you add an explanation (similar to the below) say around L3160, e.g. for a future reader to understand what this algorithm is doing.

**Comment by lukemanley** [comment]
> yep needs a rebase :->

rebased this one

File: doc/source/whatsnew/v1.5.0.rst [hunk 2]
```diff
@@ -310,7 +310,7 @@ Performance improvements
 - Performance improvement in :meth:`.GroupBy.diff` (:issue:`16706`)
 - Performance improvement in :meth:`.GroupBy.transform` when broadcasting values for user-defined functions (:issue:`45708`)
 - Performance improvement in :meth:`.GroupBy.transform` for user-defined functions when only a single group exists (:issue:`44977`)
-- Performance improvement in :meth:`MultiIndex.get_locs` (:issue:`45681`, :issue:`46040`)
+- Performance improvement in :meth:`MultiIndex.get_locs` (:issue:`45681`, :issue:`46040`, :issue:`46330`)
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v1.5.0.rst:313]: most users dont use get_locs directly; is there a more user-facing description?

Comment (lukemanley) [reply doc/source/whatsnew/v1.5.0.rst:313]: How about this:

Performance improvement in :meth:`DataFrame.loc` and :meth:`Series.loc` for tuple-based indexing of a :class:`MultiIndex`

**Comment by jbrockmendel** [comment]
LGTM

File: doc/source/whatsnew/v1.5.0.rst (hunk 2 above)
Comment (lukemanley) [reply doc/source/whatsnew/v1.5.0.rst]: updated, thanks

**Comment by jreback** [comment]
can you merge master once again

**Comment by lukemanley** [comment]
@jreback - merged main and greenish. I don't think the error is related as I see it showing up in other PRs as well

File: pandas/core/indexes/multi.py [hunk 3]
```diff
@@ -3418,82 +3383,63 @@ def _update_indexer(idxr: Index, indexer: Index | None) -> Index:
...
             else:
-                # a single label
-                lvl_indexer = self._get_loc_level(k, level=i)[0]
-                indexer = _update_indexer(
-                    _convert_to_indexer(lvl_indexer),
-                    indexer=indexer,
-                )
+                indexer &= lvl_indexer
+                if not np.any(indexer) and np.any(lvl_indexer):
+                    raise KeyError(seq)
```
Comment (jreback) [inline pandas/core/indexes/multi.py:3415]: this hit by tests?

Comment (lukemanley) [reply pandas/core/indexes/multi.py:3415]: Yes, covered by test_loc.py > test_missing_key_combination

**Comment by jreback** [comment]
thanks @lukemanley

**Pull request merged**